expression/src/visitors.rs

```diff
@@ -0,0 +1,199 @@
+use std::{iter, ops::ControlFlow};
+
+use crate::AlgebraicExpression;
+
+/// Generic trait that allows to iterate over sub-structures.
+///
+/// It is only meant to iterate non-recursively over the direct children.
+/// Self and O do not have to be the same type and we can also have
+/// Children<O1> and Children<O2> implemented for the same type,
+/// if the goal is to iterate over sub-structures of different kinds.
+pub trait Children<O> {
+    /// Returns an iterator over all direct children of kind O in this object.
+    fn children(&self) -> Box<dyn Iterator<Item = &O> + '_>;
+    /// Returns an iterator over all direct children of kind Q in this object.
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut O> + '_>;
+}
+
+pub trait AllChildren<O> {
+    /// Returns an iterator over all direct and indirect children of kind `O` in this object.
+    /// If `O` and `Self` are the same type, also includes `self`.
+    /// Pre-order visitor.
+    fn all_children(&self) -> Box<dyn Iterator<Item = &O> + '_>;
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum VisitOrder {
+    Pre,
+    Post,
+}
+
+/// A trait to be implemented by an AST node.
+///
+/// The idea is that it calls a callback function on each of the sub-nodes
+/// that are expressions.
+/// The difference to the Children<Expr> trait is that ExpressionVisitable
+/// visits recursively.
+/// If a node implements Children<Expr>, it also implements ExpressionVisitable<Expr>.
+pub trait ExpressionVisitable<Expr> {
+    /// Traverses the AST and calls `f` on each Expression in pre-order,
+    /// potentially break early and return a value.
+    fn pre_visit_expressions_return_mut<F, B>(&mut self, f: &mut F) -> ControlFlow<B>
```

Comment:
Do we actually need all of these?

Reply:
I'm not sure! Should I try to trim it here, or use the same trait in `powdr_ast`?